js/libs/builder.js

```diff
@@ -708,6 +708,15 @@ new Vue({
         $vm.triggerSave();
       });
     },
+    toggleAccessType: function(type, isTypeActive) {
+      var typeIndex = this.accessRules[0].type.indexOf(type);
```

Comment:
Why are we only looking for the first one rather than searching the entire list of access rules?

And if `this.accessRules` is empty, `this.accessRules[0].type` will result in an error.

Reply:
@tonytlwu `this.accessRules` is a default variable set by us and it's always present and always has only one element in it. Will it be changed in the future?

Reply:
I don't know if it will change in the future, but a search within the entire list will not change the output behaviour and shouldn't add much more complexity. Let'd do that.

Reply:
@tonytlwu While I was implementing this solution question appeared.
As you may see in the next lines we are using the same way to add or remove a rule type. But how we should determine to what access rules we should add or remove the new type? Should we do it for all of them in for each loop?

Reply:
@YaroslavOvdii See Slide 11

* If any of the required security rules is missing (based on active features), show the security rule warning.
* The security rules are verified by the **type** of each rule, not by **user condition**.
* When a user deactivates a feature requires a security rule:
    * Security rules are not assessed or updated or removed (by design)
        * Long term solution is to add security check before launch


And Slide 12

* Check the data source security rules to make sure the security rule warning is shown if any of the required rules (based on the active features) is missing.

---

We should always check directly if the data source has the right access rules based on the rule types. And rules don't need to be removed or updated of the activated features are changed. We just need to make sure we don't add multiple of the same type.